filecoin-proofs/src/api/internal.rs

```diff
@@ -70,6 +74,57 @@ lazy_static! {
         read_cached_params(&official_post_params_path()).ok();
 }
 
+////////////////////////////////////////////////////////////////////////////////
+/// Groth Params/Verifying-keys Memory Cache
+
+type Bls12GrothParams = groth16::Parameters<Bls12>;
+type Bls12VerifyingKey = groth16::VerifyingKey<Bls12>;
+type GrothMemCache = HashMap<String, Bls12GrothParams>;
+type VerifyingKeyMemCache = HashMap<String, Bls12VerifyingKey>;
+
+lazy_static! {
+    static ref GROTH_PARAM_MEMORY_CACHE: Mutex<GrothMemCache> = Default::default();
+    static ref VERIFYING_KEY_MEMORY_CACHE: Mutex<VerifyingKeyMemCache> = Default::default();
+}
+
+fn lookup_groth_params<F: FnOnce() -> error::Result<Bls12GrothParams>>(
+    identifier: String,
+    generator: F,
+) -> error::Result<Bls12GrothParams> {
+    let cache = &mut (*GROTH_PARAM_MEMORY_CACHE).lock().unwrap();
+    info!(FCP_LOG, "trying groth parameters memory cache for: {}", &identifier; "target" => "params");
+    let params = match cache.entry(identifier) {
+        Entry::Vacant(entry) => entry.insert(generator()?).clone(),
```

Comment:
It seems pretty inefficient to `clone` these values if we fetch them from the cache, we should be able to return a reference to them. 

Doesn't have to be in this PR, I looked into the changes involved and it is a bit involved pulling `Box`s and `&`s through, but should be worth it to reduce memory usage

Reply:
If you know how to make this work (completely), I would love to see. I tried a number of things and couldn't avoid the `clone`.

I think the root issue is that `create_proof` takes ownership of the groth params it is passed — so it seems we have to clone them prior to use.

I could easily be missing a sophistication, though. Please feel free to attack this in a follow-up.

Reply:
yeah, I have fought this problem recently somewhere else and I am pretty sure I know how to make it work: will work in on it soonish: https://github.com/filecoin-project/rust-fil-proofs/issues/499

Reply:
Looking more closely at the source, I see that we *should* be able to pass a reference to `create_proof` — so… should work out in the end. Thanks for flagging.